### PR TITLE
#162725218 Users should be able to signup

### PR DESCRIPTION
**Pull request by armstrongsouljah**
##### What does this Pr do?
- Adds an implementation to enable users to create new accounts on the authors' heaven platform.

##### Description of the task to be implemented.
- Create  a signup form to allow users to enter the details for their new accounts
- Create a component to handle the signup process by sending user information to the backend API.
- Connect the component to the global store.
- When users enter incorrect details, they should be presented with descriptive validation errors.

#### How can this be manually tested?
- Fetch this branch to your local machine by running `git fetch origin ft-users-should-signup-162725218` in your terminal.
- Checkout to this branch by running ` git checkout ft-users-should-signup-162725218` in your terminal.
- Install project dependencies by running `yarn install or just yarn`
- Start the development server by running `yarn start` in your terminal.
- Navigate to the signup route by clicking on the `Get Started` link in the browser.
- Enter your details into the form and click the signup button.
- On success, you should receive an email inviting you to verify your email.

#### What are the related pivotal stories?
[162725218](https://www.pivotaltracker.com/story/show/162725218)

### CODE REVIEW

**Comment by MuhweziDeo** [comment]
@armstrongsouljah  everything looks good, redux actions are dispatched and work fine, but maybe you would want to save the access token created on signup in local storage. such as the users can access the platform

File: src/actions/__tests__/signupAction.test.js [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+const mockStore = configureMockStore(middlewares);
+
+const postUrl = 'https://ah-backend-xmen.herokuapp.com/api/users/';
+const validData = {
+  username: 'sdsdserefefefefe',
+  email: 'maildedd@yahoo.com',
+  password: 'Password123',
+};
+
+describe('async actions', () => {
```
Comment (RrNn) [inline src/actions/__tests__/signupAction.test.js]: You need to update the description here. What are `async actions`?

File: src/actions/__tests__/signupAction.test.js [hunk 2]
```diff
@@ -0,0 +1,73 @@
+import configureMockStore from 'redux-mock-store';
+import thunk from 'redux-thunk';
+import fetchMock from 'fetch-mock';
+import registerUser from '../signupAction';
+import ActionTypes from '../ActionTypes';
+
+const middlewares = [thunk];
+const mockStore = configureMockStore(middlewares);
+
+const postUrl = 'https://ah-backend-xmen.herokuapp.com/api/users/';
```
Comment (RrNn) [inline src/actions/__tests__/signupAction.test.js]: Remember to get this URI from the environment later.

File: src/actions/__tests__/signupAction.test.js [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+      mode: 'cors',
+    });
+    const store = mockStore();
+
+    return store.dispatch(registerUser(validData)).then(() => {
+      expect(store.getActions()).toEqual(expectedAction);
+    });
+  });
+
+  it('creates REGISTER_FAILURE when registerUser is done', () => {
```
Comment (RrNn) [inline src/actions/__tests__/signupAction.test.js:50]: Perhaps it **dispatches**, not **creates**

File: src/components/Header.js [hunk 4]
```diff
@@ -1,12 +1,48 @@
-import React from 'react';
+/* eslint-disable import/no-named-as-default */
```
Comment (RrNn) [inline src/components/Header.js]: Why are you disabling our ESLint?

File: src/views/__tests__/RegisterView.test.js [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+    instance.setState = jest.fn();
+    const wrapper = mount(
+      instance.render(),
+    );
+    wrapper.find('#username').simulate('change');
+    expect(instance.handleUsernameChange).toHaveBeenCalled();
+  });
+  it('should call handleOnUsernameChange', () => {
+    const instance = new RegisterView(props);
+    // const handleUsernameChanged = jest.fn();
```
Comment (RrNn) [inline src/views/__tests__/RegisterView.test.js]: Remove the commented out code.

File: src/routes/index.js [hunk 6]
```diff
@@ -4,6 +4,7 @@ import { BrowserRouter as Router, Route } from 'react-router-dom';
 import HomeView from '../views/HomeView';
 import LoginView from '../views/LoginView';
 import Header from '../components/Header';
+// import RegisterView from '../views/RegisterView';
```
Comment (RrNn) [inline src/routes/index.js]: Remove commented out code.

File: src/views/RegisterView.js [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+
+export class RegisterView extends Component {
+  state = {
+    username: '',
+    email: '',
+    password: '',
+  }
+
+  handleErrors = (signupState) => {
+    // const { signupState } = this.props;
```
Comment (RrNn) [inline src/views/RegisterView.js]: Remove this line

File: src/views/RegisterView.js [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+    this.setState({ password: event.target.value });
+  }
+
+  handleSubmit = (event) => {
+    event.preventDefault();
+    const data = {};
+    data.username = this.state.username;
+    data.email = this.state.email;
+    data.password = this.state.password;
+    // eslint-disable-next-line no-shadow
```
Comment (RrNn) [inline src/views/RegisterView.js]: Remove this line

File: src/views/__tests__/RegisterView.test.js [hunk 9]
```diff
@@ -0,0 +1,84 @@
+/* eslint-disable no-shadow */
+/* eslint-disable prefer-const */
+import { mount } from 'enzyme';
+// import Adapter from 'enzyme-adapter-react-16';
```
Comment (RrNn) [inline src/views/__tests__/RegisterView.test.js]: Remove this line

**Review by RrNn — CHANGES_REQUESTED** [review]

File: src/actions/__tests__/signupAction.test.js (hunk 1 above)
Comment (armstrongsouljah) [reply src/actions/__tests__/signupAction.test.js]: Thank you @RrNn for pointing this out. Let me make the test more descriptive.

File: src/components/Header.js (hunk 4 above)
Comment (armstrongsouljah) [reply src/components/Header.js]: I forgot about these lines. But am looking into them.

**Review by MuhweziDeo — DISMISSED** [review]
LGTM @armstrongsouljah  everything seems to be running fine

**Review by MuhweziDeo — COMMENTED** [review]
Everything checks out

**Review by calvinpete — DISMISSED** [review]
LGTM

**Review by MuhweziDeo — DISMISSED** [review]
@armstrongsouljah  Good work on this, everything checks out

**Pull request merged**